File: go/internal/tools/todos_mcp_tool_test.go
package tools

import (
	"strings"
	"testing"
	"time"

	"github.com/matpadley/MCPServer_Demo/go/internal/data"
)

func createTestDatabase(t *testing.T) *data.DatabaseContext {
	db, err := data.NewInMemoryDatabaseContext()
	if err != nil {
		t.Fatalf("Failed to create in-memory database: %v", err)
	}
	return db
}

func TestCreateTodoAsync(t *testing.T) {
	db := createTestDatabase(t)
	defer db.Close()

	tool := NewTodosMcpTool(db)
	description := "Test todo"
	createdDate := time.Now()

	result, err := tool.CreateTodoAsync(description, createdDate)
	if err != nil {
		t.Fatalf("CreateTodoAsync failed: %v", err)
	}

	if !strings.Contains(result, "Todo created:") {
		t.Errorf("Expected success message, got: %s", result)
	}
	if !strings.Contains(result, description) {
		t.Errorf("Expected result to contain description %s, got: %s", description, result)
	}
	if !strings.Contains(result, "(Id:") {
		t.Errorf("Expected result to contain ID, got: %s", result)
	}
}

func TestReadTodosAsync_AllTodos(t *testing.T) {
	db := createTestDatabase(t)
	defer db.Close()

	tool := NewTodosMcpTool(db)

	// Create test todos
	_, _ = tool.CreateTodoAsync("Todo 1", time.Now())
	_, _ = tool.CreateTodoAsync("Todo 2", time.Now().Add(time.Hour))

	todos, err := tool.ReadTodosAsync(nil)
	if err != nil {
		t.Fatalf("ReadTodosAsync failed: %v", err)
	}

	if len(todos) != 2 {
		t.Errorf("Expected 2 todos, got %d", len(todos))
	}
}

func TestReadTodosAsync_SpecificTodo(t *testing.T) {
	db := createTestDatabase(t)
	defer db.Close()

	tool := NewTodosMcpTool(db)

	// Create test todo
	_, _ = tool.CreateTodoAsync("Test todo", time.Now())

	// Read by ID
	id := "1"
	todos, err := tool.ReadTodosAsync(&id)
	if err != nil {
		t.Fatalf("ReadTodosAsync failed: %v", err)
	}

	if len(todos) != 1 {
		t.Errorf("Expected 1 todo, got %d", len(todos))
	}
	if todos[0].ID != 1 {
		t.Errorf("Expected todo ID 1, got %d", todos[0].ID)
	}
}

func TestReadTodosAsync_InvalidId(t *testing.T) {
	db := createTestDatabase(t)
	defer db.Close()

	tool := NewTodosMcpTool(db)

	// Try to read with invalid ID
	invalidId := "invalid"
	todos, err := tool.ReadTodosAsync(&invalidId)
	if err != nil {
		t.Fatalf("ReadTodosAsync failed: %v", err)
	}

	if len(todos) != 0 {
		t.Errorf("Expected 0 todos for invalid ID, got %d", len(todos))
	}
}

func TestReadTodosAsync_NonExistentId(t *testing.T) {
	db := createTestDatabase(t)
	defer db.Close()

	tool := NewTodosMcpTool(db)

	// Try to read non-existent todo
	id := "999"
	todos, err := tool.ReadTodosAsync(&id)
	if err != nil {
		t.Fatalf("ReadTodosAsync failed: %v", err)
	}

	if len(todos) != 0 {
		t.Errorf("Expected 0 todos for non-existent ID, got %d", len(todos))
	}
}

func TestUpdateTodoAsync_ValidId(t *testing.T) {
	db := createTestDatabase(t)
	defer db.Close()

	tool := NewTodosMcpTool(db)

	// Create test todo
	_, _ = tool.CreateTodoAsync("Original description", time.Now())

	newDescription := "Updated description"
	newDate := time.Now().Add(24 * time.Hour)
	result, err := tool.UpdateTodoAsync("1", &newDescription, &newDate)
	if err != nil {
		t.Fatalf("UpdateTodoAsync failed: %v", err)
	}

	if result != "Todo 1 updated." {
		t.Errorf("Expected 'Todo 1 updated.', got: %s", result)
	}

	// Verify update
	id := "1"
	todos, _ := tool.ReadTodosAsync(&id)
	if len(todos) != 1 {
		t.Fatalf("Expected 1 todo after update, got %d", len(todos))
	}
	if todos[0].Description == nil || *todos[0].Description != newDescription {
		t.Errorf("Expected description %s, got %v", newDescription, todos[0].Description)
	}
	if !todos[0].CreatedDate.Equal(newDate) {
		t.Errorf("Expected created date %v, got %v", newDate, todos[0].CreatedDate)
	}
}

func TestUpdateTodoAsync_OnlyDescription(t *testing.T) {
	db := createTestDatabase(t)
	defer db.Close()

	tool := NewTodosMcpTool(db)

	originalDate := time.Now()
	_, _ = tool.CreateTodoAsync("Original description", originalDate)

	newDescription := "Updated description"
	result, err := tool.UpdateTodoAsync("1", &newDescription, nil)
	if err != nil {
		t.Fatalf("UpdateTodoAsync failed: %v", err)
	}

	if result != "Todo 1 updated." {
		t.Errorf("Expected 'Todo 1 updated.', got: %s", result)
	}

	// Verify update
	id := "1"
	todos, _ := tool.ReadTodosAsync(&id)
	if len(todos) != 1 {
		t.Fatalf("Expected 1 todo after update, got %d", len(todos))
	}
	if todos[0].Description == nil || *todos[0].Description != newDescription {
		t.Errorf("Expected description %s, got %v", newDescription, todos[0].Description)
	}
	if !todos[0].CreatedDate.Equal(originalDate) {
		t.Errorf("Expected original date %v to be unchanged, got %v", originalDate, todos[0].CreatedDate)
	}
}

func TestUpdateTodoAsync_EmptyDescription(t *testing.T) {
	db := createTestDatabase(t)
	defer db.Close()

	tool := NewTodosMcpTool(db)

	originalDescription := "Original description"
	_, _ = tool.CreateTodoAsync(originalDescription, time.Now())

	emptyDescription := ""
	newDate := time.Now().Add(24 * time.Hour)
	result, err := tool.UpdateTodoAsync("1", &emptyDescription, &newDate)
	if err != nil {
		t.Fatalf("UpdateTodoAsync failed: %v", err)
	}

	if result != "Todo 1 updated." {
		t.Errorf("Expected 'Todo 1 updated.', got: %s", result)
	}

	// Verify description wasn't changed
	id := "1"
	todos, _ := tool.ReadTodosAsync(&id)
	if len(todos) != 1 {
		t.Fatalf("Expected 1 todo after update, got %d", len(todos))
	}
	if todos[0].Description == nil || *todos[0].Description != originalDescription {
		t.Errorf("Expected description to remain %s, got %v", originalDescription, todos[0].Description)
	}
	if !todos[0].CreatedDate.Equal(newDate) {
		t.Errorf("Expected created date %v, got %v", newDate, todos[0].CreatedDate)
	}
}

func TestUpdateTodoAsync_InvalidId(t *testing.T) {
	db := createTestDatabase(t)
	defer db.Close()

	tool := NewTodosMcpTool(db)

	newDescription := "New description"
	result, err := tool.UpdateTodoAsync("invalid", &newDescription, nil)
	if err != nil {
		t.Fatalf("UpdateTodoAsync failed: %v", err)
	}

	if result != "Invalid todo id." {
		t.Errorf("Expected 'Invalid todo id.', got: %s", result)
	}
}

func TestUpdateTodoAsync_NonExistentId(t *testing.T) {
	db := createTestDatabase(t)
	defer db.Close()

	tool := NewTodosMcpTool(db)

	newDescription := "New description"
	result, err := tool.UpdateTodoAsync("999", &newDescription, nil)
	if err != nil {
		t.Fatalf("UpdateTodoAsync failed: %v", err)
	}

	if result != "Todo with Id 999 not found." {
		t.Errorf("Expected 'Todo with Id 999 not found.', got: %s", result)
	}
}

func TestDeleteTodoAsync_ValidId(t *testing.T) {
	db := createTestDatabase(t)
	defer db.Close()

	tool := NewTodosMcpTool(db)

	// Create test todo
	_, _ = tool.CreateTodoAsync("Test todo", time.Now())

	result, err := tool.DeleteTodoAsync("1")
	if err != nil {
		t.Fatalf("DeleteTodoAsync failed: %v", err)
	}

	if result != "Todo 1 deleted." {
		t.Errorf("Expected 'Todo 1 deleted.', got: %s", result)
	}

	// Verify deletion
	id := "1"
	todos, _ := tool.ReadTodosAsync(&id)
	if len(todos) != 0 {
		t.Errorf("Expected 0 todos after deletion, got %d", len(todos))
	}
}

func TestDeleteTodoAsync_InvalidId(t *testing.T) {
	db := createTestDatabase(t)
	defer db.Close()

	tool := NewTodosMcpTool(db)

	result, err := tool.DeleteTodoAsync("invalid")
	if err != nil {
		t.Fatalf("DeleteTodoAsync failed: %v", err)
	}

	if result != "Invalid todo id." {
		t.Errorf("Expected 'Invalid todo id.', got: %s", result)
	}
}

func TestDeleteTodoAsync_NonExistentId(t *testing.T) {
	db := createTestDatabase(t)
	defer db.Close()

	tool := NewTodosMcpTool(db)

	result, err := tool.DeleteTodoAsync("999")
	if err != nil {
		t.Fatalf("DeleteTodoAsync failed: %v", err)
	}

	if result != "Todo with Id 999 not found." {
		t.Errorf("Expected 'Todo with Id 999 not found.', got: %s", result)
	}
}

func TestMultipleOperationsInSequence(t *testing.T) {
	db := createTestDatabase(t)
	defer db.Close()

	tool := NewTodosMcpTool(db)

	// Create multiple todos
	_, _ = tool.CreateTodoAsync("Todo 1", time.Now())
	_, _ = tool.CreateTodoAsync("Todo 2", time.Now().Add(time.Hour))
	_, _ = tool.CreateTodoAsync("Todo 3", time.Now().Add(2*time.Hour))

	// Read all todos
	todos, _ := tool.ReadTodosAsync(nil)
	if len(todos) != 3 {
		t.Errorf("Expected 3 todos, got %d", len(todos))
	}

	// Update middle todo
	newDescription := "Updated Todo 2"
	_, _ = tool.UpdateTodoAsync("2", &newDescription, nil)

	// Delete first todo
	_, _ = tool.DeleteTodoAsync("1")

	// Verify final state
	todos, _ = tool.ReadTodosAsync(nil)
	if len(todos) != 2 {
		t.Errorf("Expected 2 todos after operations, got %d", len(todos))
	}

	// Check that todo 2 was updated and todo 1 is gone
	foundUpdatedTodo := false
	for _, todo := range todos {
		if todo.ID == 2 && todo.Description != nil && *todo.Description == newDescription {
			foundUpdatedTodo = true
		}
		if todo.ID == 1 {
			t.Error("Todo 1 should have been deleted")
		}
	}
	if !foundUpdatedTodo {
		t.Error("Todo 2 should have been updated")
	}
}

File: typescript/tests/TodosMcpTool.test.ts
import { TodosMcpTool } from '../src/tools/TodosMcpTool';
import { DatabaseContext } from '../src/data/DatabaseContext';
import { Todo } from '../src/data/Todo';
import * as fs from 'fs';
import * as path from 'path';

/**
 * Comprehensive tests for TodosMcpTool class
 * Tests all CRUD operations and error handling scenarios
 */
describe('TodosMcpTool', () => {
  let todosTool: TodosMcpTool;
  let dbContext: DatabaseContext;
  let testDbPath: string;

  beforeEach(async () => {
    // Create a unique test database for each test
    testDbPath = path.join(__dirname, `test-${Date.now()}-${Math.random()}.db`);
    dbContext = new DatabaseContext(testDbPath);
    // Wait a bit for database initialization
    await new Promise(resolve => setTimeout(resolve, 100));
    todosTool = new TodosMcpTool(dbContext);
  });

  afterEach(async () => {
    // Clean up test database
    dbContext.close();
    if (fs.existsSync(testDbPath)) {
      fs.unlinkSync(testDbPath);
    }
  });

  describe('createTodoAsync', () => {
    it('should create a todo successfully', async () => {
      // Arrange
      const description = 'Test todo';
      const createdDate = new Date('2024-01-01');

      // Act
      const result = await todosTool.createTodoAsync(description, createdDate);

      // Assert
      expect(result).toContain('Todo created:');
      expect(result).toContain(description);
      
      const todos = await todosTool.readTodosAsync();
      expect(todos).toHaveLength(1);
      expect(todos[0].description).toBe(description);
      expect(todos[0].createdDate).toEqual(createdDate);
    });
  });

  describe('readTodosAsync', () => {
    beforeEach(async () => {
      // Add test data
      await todosTool.createTodoAsync('Todo 1', new Date('2024-01-01'));
      await todosTool.createTodoAsync('Todo 2', new Date('2024-01-02'));
    });

    it('should return all todos when no id is provided', async () => {
      // Act
      const todos = await todosTool.readTodosAsync();

      // Assert
      expect(todos).toHaveLength(2);
      expect(todos[0].description).toBe('Todo 1');
      expect(todos[1].description).toBe('Todo 2');
    });

    it('should return all todos when empty string id is provided', async () => {
      // Act
      const todos = await todosTool.readTodosAsync('');

      // Assert
      expect(todos).toHaveLength(2);
    });

    it('should return specific todo when valid id is provided', async () => {
      // Arrange
      const allTodos = await todosTool.readTodosAsync();
      const firstTodoId = allTodos[0].id;

      // Act
      const todos = await todosTool.readTodosAsync(firstTodoId.toString());

      // Assert
      expect(todos).toHaveLength(1);
      expect(todos[0].id).toBe(firstTodoId);
      expect(todos[0].description).toBe('Todo 1');
    });

    it('should return empty array when invalid numeric id is provided', async () => {
      // Act
      const todos = await todosTool.readTodosAsync('999');

      // Assert
      expect(todos).toHaveLength(0);
    });

    it('should return empty array when non-numeric id is provided', async () => {
      // Act
      const todos = await todosTool.readTodosAsync('invalid');

      // Assert
      expect(todos).toHaveLength(0);
    });
  });

  describe('updateTodoAsync', () => {
    let todoId: number;

    beforeEach(async () => {
      await todosTool.createTodoAsync('Original todo', new Date('2024-01-01'));
      const todos = await todosTool.readTodosAsync();
      todoId = todos[0].id;
    });

    it('should update todo description and date successfully', async () => {
      // Arrange
      const newDescription = 'Updated todo';
      const newDate = new Date('2024-01-02');

      // Act
      const result = await todosTool.updateTodoAsync(todoId.toString(), newDescription, newDate);

      // Assert
      expect(result).toBe(`Todo ${todoId} updated.`);
      
      const todos = await todosTool.readTodosAsync(todoId.toString());
      expect(todos[0].description).toBe(newDescription);
      expect(todos[0].createdDate).toEqual(newDate);
    });

    it('should update only description when date is not provided', async () => {
      // Arrange
      const newDescription = 'Updated description';
      const originalDate = new Date('2024-01-01');

      // Act
      const result = await todosTool.updateTodoAsync(todoId.toString(), newDescription);

      // Assert
      expect(result).toBe(`Todo ${todoId} updated.`);
      
      const todos = await todosTool.readTodosAsync(todoId.toString());
      expect(todos[0].description).toBe(newDescription);
      expect(todos[0].createdDate).toEqual(originalDate);
    });

    it('should not update description when empty string is provided', async () => {
      // Arrange
      const newDate = new Date('2024-01-02');

      // Act
      const result = await todosTool.updateTodoAsync(todoId.toString(), '', newDate);

      // Assert
      expect(result).toBe(`Todo ${todoId} updated.`);
      
      const todos = await todosTool.readTodosAsync(todoId.toString());
      expect(todos[0].description).toBe('Original todo'); // Unchanged
      expect(todos[0].createdDate).toEqual(newDate);
    });

    it('should return error for invalid id', async () => {
      // Act
      const result = await todosTool.updateTodoAsync('invalid', 'New description');

      // Assert
      expect(result).toBe('Invalid todo id.');
    });

    it('should return error for non-existent id', async () => {
      // Act
      const result = await todosTool.updateTodoAsync('999', 'New description');

      // Assert
      expect(result).toBe('Todo with Id 999 not found.');
    });
  });

  describe('deleteTodoAsync', () => {
    let todoId: number;

    beforeEach(async () => {
      await todosTool.createTodoAsync('To be deleted', new Date());
      const todos = await todosTool.readTodosAsync();
      todoId = todos[0].id;
    });

    it('should delete todo successfully', async () => {
      // Act
      const result = await todosTool.deleteTodoAsync(todoId.toString());

      // Assert
      expect(result).toBe(`Todo ${todoId} deleted.`);
      
      const todos = await todosTool.readTodosAsync();
      expect(todos).toHaveLength(0);
    });

    it('should return error for invalid id', async () => {
      // Act
      const result = await todosTool.deleteTodoAsync('invalid');

      // Assert
      expect(result).toBe('Invalid todo id.');
    });

    it('should return error for non-existent id', async () => {
      // Act
      const result = await todosTool.deleteTodoAsync('999');

      // Assert
      expect(result).toBe('Todo with Id 999 not found.');
    });
  });

  describe('edge cases and comprehensive scenarios', () => {
    it('should handle multiple operations in sequence', async () => {
      // Create multiple todos
      await todosTool.createTodoAsync('Todo 1', new Date('2024-01-01'));
      await todosTool.createTodoAsync('Todo 2', new Date('2024-01-02'));
      await todosTool.createTodoAsync('Todo 3', new Date('2024-01-03'));

      // Read all todos
      let todos = await todosTool.readTodosAsync();
      expect(todos).toHaveLength(3);

      // Update middle todo
      const middleTodoId = todos[1].id;
      await todosTool.updateTodoAsync(middleTodoId.toString(), 'Updated Todo 2');

      // Delete first todo
      const firstTodoId = todos[0].id;
      await todosTool.deleteTodoAsync(firstTodoId.toString());

      // Verify final state
      todos = await todosTool.readTodosAsync();
      expect(todos).toHaveLength(2);
      expect(todos.find(t => t.id === middleTodoId)?.description).toBe('Updated Todo 2');
      expect(todos.find(t => t.id === firstTodoId)).toBeUndefined();
    });

    it('should handle todos with same descriptions but different dates', async () => {
      // Arrange
      const description = 'Same description';
      const date1 = new Date('2024-01-01');
      const date2 = new Date('2024-01-02');

      // Act
      await todosTool.createTodoAsync(description, date1);
      await todosTool.createTodoAsync(description, date2);

      // Assert
      const todos = await todosTool.readTodosAsync();
      expect(todos).toHaveLength(2);
      expect(todos[0].description).toBe(description);
      expect(todos[1].description).toBe(description);
      expect(todos[0].createdDate).toEqual(date1);
      expect(todos[1].createdDate).toEqual(date2);
    });

    it('should preserve id assignment and ordering', async () => {
      // Create todos in sequence
      await todosTool.createTodoAsync('First', new Date());
      await todosTool.createTodoAsync('Second', new Date());
      await todosTool.createTodoAsync('Third', new Date());

      const todos = await todosTool.readTodosAsync();
      
      // IDs should be sequential
      expect(todos[0].id).toBe(1);
      expect(todos[1].id).toBe(2);
      expect(todos[2].id).toBe(3);
      
      // Should be ordered by ID
      expect(todos[0].description).toBe('First');
      expect(todos[1].description).toBe('Second');
      expect(todos[2].description).toBe('Third');
    });
  });
});